proposals/testing/NNNN-targeted-interoperability-swift-testing-and-xctest.md

```diff
@@ -0,0 +1,292 @@
+
+## Motivation
+
+Unfortunately, mixing an API call from one framework with a test from the other
+framework may not work as expected. As a more concrete example, if you take an
+existing test helper function written for XCTest and call it in a Swift Testing
+test, it won't report the assertion failure:
+
+```swift
+func assertUnique(_ elements: [Int]) {
+  if Set(elements).count != elements.count {
+    XCTFail("\(elements) has non unique elements")
+  }
+}
+
+// XCTest
+
+class FooTests: XCTestCase {
+    func testDups() {
+        assertUnique([1, 2, 1]) // Fails as expected
+    }
+}
+
+// Swift Testing
+
+@Test func `Duplicate elements`() {
+  assertUnique([1, 2, 1]) // Passes? Oh no!
+}
+```
+
+Generally, we get into trouble today when ALL the following conditions are met:
+
+- You call XCTest API in a Swift Testing test, or call Swift Testing API in a
+  XCTest test
+- The API doesn't function as expected in some or all cases, and
+- You get no notice at build time or runtime about the malfunction
+
+For the remainder of this proposal, we’ll describe tests which exhibit this
+problem as **lossy without interop**.
+
+If you've switched completely to Swift Testing and don't expect to use XCTest in
+the future, this proposal includes a mechanism to **prevent you from
+inadvertently introducing XCTest APIs to your project**, including via a testing
+library.
+
+## Proposed solution
+
+- **XCTest APIs which are lossy without interop will work as expected when
+  called in Swift Testing.** In addition, runtime diagnostics will be included
+  to highlight opportunities to adopt newer Swift Testing equivalent APIs.
+
+- Conversely, **Swift Testing API called in XCTest will work as expected if
+  XCTest already provides similar functionality.** You should always feel
+  empowered to choose Swift Testing when writing new tests or test helpers, as
+  it will work properly in both types of tests.
+
+We don't propose supporting interoperability for APIs without risk of data loss,
+because they naturally have high visibility. For example, using `throw XCTSkip`
+in a Swift Testing test results in a test failure rather than a test skip,
+providing a clear indication that migration is needed.
+
+## Detailed design
+
+### Highlight and support XCTest APIs which are lossy without interop
+
+We propose supporting the following XCTest APIs in Swift Testing:
+
+- Assertions: `XCTAssert*` and [unconditional failure][] `XCTFail`
+- [Expected failures][], such as `XCTExpectFailure`: marking a Swift Testing
+  issue in this way will generate a runtime warning issue.
+- `XCTAttachment`
+- [Issue handling traits][]: we will make our best effort to translate issues
+  from XCTest to Swift Testing. Note that there are certain issue kinds that are
+  new to Swift Testing and not expressible from XCTest.
+
+Note that no changes are proposed for the `XCTSkip` API, because they already
+feature prominently as a test failure to be corrected when thrown in Swift
+Testing.
```

Comment:
```suggestion
Note that no changes are proposed for the `XCTSkip` API, because they already
feature prominently as a test failure when thrown by a Swift Testing test.
```